pkg/daemon/on_disk_validation.go

```diff
@@ -5,18 +5,119 @@ import (
 	ign3types "github.com/coreos/ignition/v2/config/v3_4/types"
+	"github.com/coreos/vcontext/path"
 	"github.com/google/go-cmp/cmp"
 	mcfgv1 "github.com/openshift/api/machineconfiguration/v1"
 	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	"github.com/openshift/machine-config-operator/pkg/daemon/constants"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/klog/v2"
 )
 
+// Match the file captured in the event -> correspoding MC spec
+func validateOnDiskStateForEvent(currentConfig *mcfgv1.MachineConfig, rootPath string, path path.ContextPath) error {
+	ignconfigi, err := ctrlcommon.IgnParseWrapper(currentConfig.Spec.Config.Raw)
+	if err != nil {
+		return fmt.Errorf("failed to parse Ignition for validation: %w", err)
+	}
+	switch typedConfig := ignconfigi.(type) {
+	case ign3types.Config:
+		value, err := getValueInMCForPath(ignconfigi.(ign3types.Config), path)
+		if err != nil {
+			return fmt.Errorf("failed to get valye: %w", err)
+		}
+		switch path.Tag {
+		case "passwd":
+			v, ok := value.([]ign3types.SSHAuthorizedKey)
+			if !ok {
+				return fmt.Errorf("failed to convert")
+			}
+			if err := checkV3Passwd(v, rootPath+constants.RHCOS9SSHKeyPath); err != nil {
+				return &sshConfigDriftErr{err}
+			}
+		case "files":
+			v, ok := value.(ign3types.File)
+			if !ok {
+				return fmt.Errorf("failed to convert")
+			}
+			if err := checkV3File(v); err != nil {
+				return &fileConfigDriftErr{err}
+			}
+
+		case "units":
+			v, ok := value.(ign3types.Unit)
+			if !ok {
+				return fmt.Errorf("failed to convert")
+			}
+			if err := checkV3Unit(v, rootPath); err != nil {
+				return &unitConfigDriftErr{err}
+			}
+
+		case "dropins":
+			unitPath, err := getMCUnitPathFromDropinPath(path)
+			if err != nil {
+				return fmt.Errorf("failed to get valye: %w", err)
+			}
+			unitValue, err := getValueInMCForPath(ignconfigi.(ign3types.Config), unitPath)
+			if err != nil {
+				return fmt.Errorf("failed to get valye: %w", err)
+			}
+			unit, ok := unitValue.(ign3types.Unit)
+			if !ok {
+				return fmt.Errorf("failed to convert")
+			}
+
+			v, ok := value.(ign3types.Dropin)
+			if !ok {
+				return fmt.Errorf("failed to convert")
+			}
+			if err := checkV3Dropin(rootPath, unit, v); err != nil {
+				return &unitConfigDriftErr{err}
+			}
+
+		default:
+			return fmt.Errorf("unknown tag")
+
+		}
+		return nil
+	default:
+		return fmt.Errorf("unexpected type for ignition config: %v", typedConfig)
+	}
+}
+
+func getMCUnitPathFromDropinPath(dropinPath path.ContextPath) (path.ContextPath, error) {
+	if dropinPath.Len() < 2 {
+		return path.New(""), fmt.Errorf("dropinPath:%s is too short ", dropinPath.Path)
+	}
+	return dropinPath.Pop().Pop(), nil
+}
+
+func getValueInMCForPath(config ign3types.Config, path path.ContextPath) (interface{}, error) {
+	val := reflect.ValueOf(config)
+	for _, p := range path.Path {
+		switch v := p.(type) {
+		case string:
+			val = val.FieldByName(v)
```

Comment:
thought: I'm not very well-versed in Golang reflection and acknowledge that its an area where I need to get a bit more comfortable. Could you include some comments which explains what this does and why reflection was the best way to do it?

As an aside, we should be sure that this function has adequate test coverage by examining the test coverage report generated by the CI system.

Reply:
The `HashMap(file path on disk, path of item in MC)` created by `getFilePathsFromMachineConfig()` records the location of the spec in MC. Hence when a `fsnotify` event occurs, the HashMap needs to be looked up by its key which is the file path of fsnotify event. It iterates through the MC path (path.ContextPath is just an array of strings), and traverses the MC dynamically accessing the value of each object till the exact value needed is reached

Reply:
Yes ensured that the testing covered most of the functionality of the function